### PR TITLE
nv3: Immediate Emails and Name Change

### PR DESCRIPTION
**Pull request by zakhap**
## Description
This PR handles a few different PRs and Issues: 
- Renames Subscriptions Page to Notifications Page #148
  - Subscriptions became Notifications
  - What was previous "notifications" was a page of all a user's notification, a page view of the notifications dropdown menu that Jake wrote way back when testing notifications. I renamed it "notifications-list". 
- Updates Subscriptions Model to include parameters for email notifications #149
  - Along side, two new routes were introduced, `/enableImmediateInvites` and `/disableImmediateInvites`, which were fully based on the `/enable(disable)Subscriptions` routes.
    - Unit tests were written for these routes.
- Generates stub email notifications for immediate subscriptions #153
  - Extending `Susbcriptions.emitNotifications()`, there is a check to see if the subscription has immediate emails turned on, and if so, an email is sent.
- Some UI was 'refreshed' for construct ui, but the issue for a full construct-ui notifications page redesign has _not_ been implemented in this PR (issue with filter came up in the eng. channel, will be attended to elsewhere). 

To clarify, the Subscription model has been extended with a boolean property `.immediate_email` which if true, will trigger an email to be sent to the user. The feature for a batched unread-notification email will be a different PR and will extend the User Model, not the Subscription Model. 

## How Has This Been Tested?
I wrote tests for the routes `/enableImmediateInvites` and `/disableImmediateInvites` but I have not written tests for the actual sending of emails.

#### However, they can be tested in QA: 
- Run Migration
- log in with multiple accounts for cross-account-notifications
- Make a thread with one account
- Open SideBar menu to see 'Manage Notifications' above 'Home' button. Click 'Manage Notifications' to go the newly renamed 'Notifications Page'. 
- Select 'Active Subscriptions' and see a list of all active subscriptions.
- Hover over email button, see Tooltip, click to enable/disable immediate emails for subscription notifications.
  - Enable immediate emails for the newest thread you made.
- Go back to thread from other account and comment. 
  - See that email was sent in your inbox (in development, the function emails `test@commonwealth.im` from `emails.ts` just to prevent accidentally emailing a user about a notification on localhost. This can be changed to your email if you (person QA'ing) doesn't get the forwarded emails from the test email.). 

## Clubhouse tickets/Github issues (if appropriate):
- #148 
- #149 
- #153 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] My change should be included in the release notes.
- [ ] I have updated the documentation accordingly.
- [ ] I have linted the code locally prior to submission.

### CODE REVIEW

File: server/models/subscription.ts [hunk 1]
```diff
@@ -12,6 +17,7 @@ module.exports = (sequelize, DataTypes) => {
     category_id: { type: DataTypes.STRING, allowNull: false },
     object_id: { type: DataTypes.STRING, allowNull: false },
     is_active: { type: DataTypes.BOOLEAN, defaultValue: true, allowNull: false },
+    immediate_email: { type: DataTypes.BOOLEAN, defaultValue: false, allowNull: false },
```
Comment (dillchen) [inline server/models/subscription.ts:20]: This seems a bit weird to directly have this boolean here, even in the short term. We'll be extending this with the other options (daily, weekly, monthly, etc etc), might as well make the change now

File: client/scripts/models/NotificationSubscription.ts [hunk 2]
```diff
@@ -6,17 +6,23 @@ class NotificationSubscription {
   public readonly objectId: string;
   public readonly createdAt: moment.Moment;
 
+  private _immediateEmail: boolean;
+  public get immediateEmail() { return this._immediateEmail; }
+  public enableImmediateEmail() { this._immediateEmail = true; }
+  public disableImmediateEmail() { this._immediateEmail = false; }
```
Comment (dillchen) [inline client/scripts/models/NotificationSubscription.ts:12]: "This seems a bit weird to directly have this boolean here, even in the short term. We'll be extending this with the other options (daily, weekly, monthly, etc etc), might as well make the change now." Extends from the choice of the Subscription Model

File: client/scripts/views/components/header.ts [hunk 3]
```diff
@@ -105,6 +105,13 @@ const Header: m.Component<{}> = {
             iconLeft: Icons.SETTINGS,
             label: 'Settings'
           }),
+          (app.vm.activeAccount
+            || m.route.get() === '/settings')
+            && m(MenuItem, {
+              onclick: () => m.route.set('/notifications'),
+              iconLeft: Icons.BELL,
+              label: 'Manage Notifications'
+            }),
```
Comment (dillchen) [inline client/scripts/views/components/header.ts:114]: @gdjohnson @zakhap this will need to be added to the psuedoheader that Graham implemented

**Comment by dillchen** [comment]
- [ ] Going from within a community to the notifications manager should bring you directly to that community
- [ ] I believe these should be switches instead of buttons.

image

- [ ] It looks like in the file rename, @jnaviask code for individually managing chain notifications for a chain has been removed. This should be included in this PR, when one clicks into the "Chain"

image

**Notifications Bar**
- [ ] Notifications Bar no longer turns red to notify a user of a new notification
- [ ] New Notifications no longer highlighted in the drop down menu
- [ ] There are many duplicate notifications once one is triggered

image

I couldn't get email notifications to work, but will try again

- [ ] On trying to enable email notifications, if a user does not have an email, we should probably direct you to the settings page and allow you to update email. (I was able to "enable" email notifications for a user whom I had only logged in with an address)

===
**Enhancements**
- [ ] Would prefer to style the notifications as cards. Notifications rows can it be changed…

This whole flow is still a bit confusing imo, will need to think about this.

File: client/scripts/views/components/header.ts (hunk 3 above)
Comment (dillchen) [reply client/scripts/views/components/header.ts:114]: @zakhap can you pull in the latest changes to raymond.construct and add the manage notifications section to the pseudoheader

Comment (zakhap) [reply client/scripts/views/components/header.ts:114]: Yes I can

**Comment by zakhap** [comment]
After cleaning up some of @dillchen's comments and contextualizing this PR with him on the phone, I believe this is ready to be merged into `raymond.construct`. I moved the big comment above into the [Notifications V3 Notion](https://www.notion.so/hicommonwealth/Notification-V3-e7086804763845c4a99c2fa8a4de7c70) and will be addressing all of the UI and user experience stuff in another PR.

**Comment by drewstone** [comment]
@zakhap there are merge conflicts to be resolved here.

**Comment by dillchen** [comment]
I'm having some trouble running this branch using the prod DB. Will take a look in the morning

**Comment by dillchen** [comment]
@zakhap can you merge `raymond.construct` into here so I can continue QA'ing

**Comment by dillchen** [comment]
- [ ]  Within the manage notifications dashboard, Clicking into a Community and then into a Chain will keep the chain highlighted. It will continue to append the page with the corresponding notifications to be flipped. I believe this should be addressed within this PR.

**Comment by zakhap** [comment]
> I was unable to run `yarn reset-server` on this branch. @zakhap have you run into this?

I just fixed this bug, not longer an issue.

**Comment by zakhap** [comment]
> * [ ]  Clicking the "turn off notifications" button from within a community didn't turn off chain notifications

@dillchen this button was originally the 'new thread' subscription toggle for the community/chain level, that's what it still does. I fixed the copy to read "turn on/off new thread notifications", but in the future PR where we extend subscriptions to have an associated community/chain, I can add the every comm/chain subscription-toggle to the dropdown as well.

**Comment by gdjohnson** [comment]
Gonna take a while to work through this, but a quick design note that may not be necessary to fix here: the way the UI is set up for the NotificationManager page, I was convinced that I was supposed to be able to see my notifications and couldn't. Having just the buttons for clearing/marking as read notifications that weren't rendered was a lil disconcerting.

**Comment by gdjohnson** [comment]
What's the role of the subscription.ts page, if we're using notifications.ts? And why give the subscription object id, instead of the thread title? Feel like if I were a user, I wouldn't have the context necessary to decide what I wanted to pause or unpause, since I wouldn't know what any of the objectIds referred to

**Comment by dillchen** [comment]
> > * [ ]   Clicking the "turn off notifications" button from within a community didn't turn off chain notifications
> 
> @dillchen this button was originally the 'new thread' subscription toggle for the community/chain level, so that's what it still does. I fixed the copy to read "turn on/off new thread notifications", but in the future PR where we extend subscriptions to have an associated community/chain, I can add the every comm/chain subscription-toggle to the dropdown as well.

Could you add this to the open questions sections in Notion? Would be good to note this

**Comment by zakhap** [comment]
> * What's the role of the subscription.ts page, if we're using notifications.ts? And why give the subscription object id, instead of the thread title? Feel like if I were a user, I wouldn't have the context necessary to decide what I wanted to pause or unpause, since I wouldn't know what any of the objectIds referred to
> * Could be nice if the email button was greyed out & on hover instructed users to add an address to configure email notifications

@gdjohnson: I add these to the ongoing list of open questions for designing the subscriptions page. UI is still going to be addressed in another PR after there is more clarity on function/usability  for this NotificationSubscription Page. 
Also, this pr renames `subscription.ts` to `notifications.ts` and moves `notifications.ts` into `notifications-list.ts` which is still up for the chopping block in the redesign. 

> Could you add this to the open questions sections in Notion? Would be good to note this
@dillchen added! 

The Subscription Page "Forever" Spec with open questions at the bottom is here : https://www.notion.so/hicommonwealth/Subscriptions-Page-Forever-Spec-cd4b4257721b449396de09d9ba8716b3

**Comment by dillchen** [comment]
I think this should be okay after this last pass from me. @drewstone on to you

**Comment by drewstone** [comment]
What's the status of the test fixes @zakhap ?

**Comment by zakhap** [comment]
> What's the status of the test fixes @zakhap ?

They should all be fine except for the 3 from `raymond.construct` that I fixed over on the `zak.tests` branch/PR

**Comment by zakhap** [comment]
@drewstone regarding coverage on the subscription related routes: 

Screen Shot 2020-05-28 at 2 54 13 PM

**Comment by drewstone** [comment]
@zakhap the description for many of the tests here does not reflect our org structure, would be nice to change them today.

**Pull request merged**